GeometryCore/src/Polygon.cpp

```diff
@@ -58,13 +72,47 @@ namespace geom_utils
     // Calculate and return convex hull for the polygon
     Polygon Polygon::convexHull() const 
     { 
-        return {}; 
+        if (points.size() <= 3) return *this; //can't bild a polygon from 2 points or less
+        Polygon convexHull;
+        auto pos_of_the_most_right_point = min_element(points.begin(), points.end(), [](auto a, auto b) {return b.x > a.x; }) - points.begin();
+        int pos_of_cur_point = pos_of_the_most_right_point;
+
+        do
+        {
+            convexHull.points.push_back(points[pos_of_cur_point]);
+            int pos_of_next_point = (pos_of_cur_point + 1) % points.size();
+
+            for (int i = 0; i < points.size(); i++)
+            {   //find the most counter-clockwise point
+                Polygon tmp_poly({ points[pos_of_cur_point], points[i], points[pos_of_next_point] });
+                if (tmp_poly.orientation())
+                    pos_of_next_point = i;
+            }
+            pos_of_cur_point = pos_of_next_point;
+
+        } while (pos_of_cur_point != pos_of_the_most_right_point);
+
+        return convexHull;
     }
 
     // Return true if polygon contour is convex, false if non-convex
     bool Polygon::isConvexHull() const
-    { 
-        return {}; 
+    {
+        if (points.size() < 3) return false;
+        bool isConvex = true;
+        bool isCounterClockwise = this->orientation();
+
+        #pragma omp parallel for
+        for (int i = 0; i < points.size(); i++)
+        {
+            FPoint2D p(points[(i + 1) % points.size()].x - points[i].x, points[(i + 1) % points.size()].y - points[i].y);
+            FPoint2D p2(points[(i + 2) % points.size()].x - points[(i + 1) % points.size()].x, points[(i + 2) % points.size()].y - points[(i + 1) % points.size()].y);
+            float res = geom_utils::cross(p, p2);
+
+            if ((isCounterClockwise == true && res < 0) || (isCounterClockwise == false && res > 0)) 
+                isConvex = false;
+        }
+        return isConvex;
```

Comment:
Good... but still please see some of my suggestions:
1. `if (points.size() < 3) return false;` technically, polygon with 2 points is convex hull, because it doen't have any interior angle bigger than 180)
2. I believe, we can remove `omp` directives here and still have the same good performance. This will allow us to perform early exit from the loop, while the omp loop should carry out all iterations (omp doesn't allow returning from parallel loop). 
Please consider my suggestion below:
```
        bool ccw = orientation();
        for (size_t i = 0; i < points.size(); i++)
        {
            auto vec1 = points[(i + 1) % points.size()] - points[i];
            auto vec2 = points[(i + 2) % points.size()] - points[i];
            if (ccw ? cross(vec1, vec2) < 0 : cross(vec1, vec2) > 0)
                return false;
        }
        return true;
```

Reply:
i was trying to do it with openMP, so i had to add `isConvex` field, and perform all the iterations.
Ok, i will implement it

Reply:
> i was trying to do it with openMP, so i had to add `isConvex` field, and perform all the iterations.

Yes, thats understandable, and you are right) I just suggested approach without using opemp, which shouldn't loose in performance)

Reply:
Yes, in this case even code has better look)